### PR TITLE
Fix Navigator.pop for named routes.

### PR DESCRIPTION
**Pull request by gspencergoog**
This fixes a problem with Navigator.pop where the routes were created with MaterialPageRoute\<Null\>, which breaks any call to Navigator.pop with an argument other than Null.

### CODE REVIEW

File: packages/flutter/test/material/app_test.dart [hunk 1]
```diff
@@ -25,6 +25,23 @@ class StateMarkerState extends State<StateMarker> {
   }
 }
 
+class PopWidget extends StatelessWidget {
+  final dynamic value;
+  final String name;
+  final BuildContext context;
+
+  const PopWidget(this.context, this.name, this.value);
```
Comment (Hixie) [inline packages/flutter/test/material/app_test.dart]: nit: constructor always comes first

Comment (Hixie) [inline packages/flutter/test/material/app_test.dart]: passing a context to a widget is pretty crazy. Maybe instead of a PopWidget, use a Builder?

Comment (gspencergoog) [inline packages/flutter/test/material/app_test.dart]: Thanks.

Comment (gspencergoog) [inline packages/flutter/test/material/app_test.dart]: Yeah, I thought it was pretty crazy too, I was just looking for a way to delay the pop until after building was done.  Thanks for the suggestion: it's much simpler now.

File: examples/flutter_gallery/flutter_gallery.iml [hunk 2]
```diff
@@ -9,6 +9,7 @@
       <excludeFolder url="file://$MODULE_DIR$/packages" />
       <excludeFolder url="file://$MODULE_DIR$/test/calculator/packages" />
       <excludeFolder url="file://$MODULE_DIR$/test/packages" />
+      <excludeFolder url="file://$MODULE_DIR$/tool/packages" />
```
Comment (Hixie) [inline examples/flutter_gallery/flutter_gallery.iml]: did you mean to do this?

File: packages/flutter/flutter.iml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<module type="WEB_MODULE" version="4">
+<module type="FLUTTER_MODULE_TYPE" version="4">
```
Comment (Hixie) [inline packages/flutter/flutter.iml]: or this?

File: examples/flutter_gallery/flutter_gallery.iml (hunk 2 above)
Comment (gspencergoog) [inline examples/flutter_gallery/flutter_gallery.iml]: Urg.  No.

**Comment by Hixie** [comment]
It's a pity that there's no way to dynamically fill the type there, but:
LGTM

**Comment by gspencergoog** [comment]
Well, but where would it get the type from?: we don't know the type until pop is called, so it kind of has to be dynamic.

Anyhow, thanks for the review!

**Comment by Hixie** [comment]
The route knows the type when it's created. A language with good RTTI could let us do it. But yeah, not in Dart. :-)

**Comment by gspencergoog** [comment]
I did try this, and it seems to work:
```dart
import 'dart:async';

class Route<T> {
  /// Will throw if arg isn't the right type.
  bool _isCorrectType(T arg) => true;
}

class Bar {}

class Navigator {
  Completer completer = new Completer();
  Route<dynamic> route;

  Future<dynamic> push(Route<dynamic> newRoute) {
    print("Pushing ${newRoute}");
    route = newRoute;
    return completer.future;
  }

  void pop(dynamic result) {
    print("Popping with ${result} (${result.runtimeType})");
    assert(route._isCorrectType(result));
    completer.complete(result);
  }
}

Future<Null> main() async {
  Navigator foo = new Navigator();
  var out1 = foo.push(new Route<String>());
  foo.pop("all done"); // Works
  print("Popped value is ${await out1}");

  var out2 = foo.push(new Route<String>());
  foo.pop(new Bar()); // Fails.
  print("Popped value is ${await out2}");
}
```

Hey, what's the deal with Travis?  Seems like most of my runs for this PR have been failing through no fault of my code...

**Comment by Hixie** [comment]
The pop doesn't determine the type. The type is there to ensure the pop is providing the right type. The type is decided by the person who defined the route. For example, a page where you edit a contact might be defined as returning a `Contact` class. It would be an error if any of the ways to pop that page returned, say, an integer or a string.

Not sure what's up with Travis.

**Comment by Hixie** [comment]
Did you mean to add the `arguments` feature?

**Comment by gspencergoog** [comment]
Dammmit.  No.  I was trying to bring my branch up to current flutter/master.  I went through my local master, which must have had that in it still.  From now on, I'm nuking my local repos before each change. :-)

Removed.

**Comment by gspencergoog** [comment]
Oh, I see what you mean: you want to validate that the Route widget created by the builder in the `routes` table matches the type we put into the MaterialPageRoute's generic type.  Yeah, no way to do that, really.

**Comment by gspencergoog** [comment]
Yes, the Route<T>.didPop validates the type, but in the case of the `routes:` map in the app, the first chance the user has to supply a type is in the pop, since the app is creating the MaterialPageRoute for them.

**Comment by Hixie** [comment]
Yeah. Like I said, it's a pity. :-) Not much I can think to do about it.

Anyway, all is green, so I've landed your patch. Thanks!

**Pull request merged**